c8d/load: Don't ignore missing platform when requested

vvoland opened the pull request:
- replaces, closes https://github.com/moby/moby/pull/48590
- fixes https://github.com/moby/moby/issues/48589
- relates to https://github.com/docker/cli/pull/5331
- relates to https://github.com/moby/moby/pull/48295

**- What I did**
Improve `docker load` with `--platform` error when loading the requested platform failed and make it error out when the requested platform is not available after the load.

**- How to verify it**
`TestImageLoadMissing`

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
containerd image store: Make `docker load --platform` return an error when the requested platform wasn't loaded.
```

**- A picture of a cute animal (not mandatory but encouraged)**

![sleepy](https://github.com/user-attachments/assets/ae03bab6-eb45-4b35-977e-faeca74d4467)

vvoland committed:
c8d/list: Extract fake service helpers

Extract to a separate file as they're also used by other files in this
package.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/load: Don't ignore missing platform when requested

Commit f143f4ec5148e71a10d074ed275af8f63caff339 introduced platform support
when loading images. However when loading a specific platform variant from
a tar that contains multiple, we should not ignore cases if that platform is
missing.

Before this patch, the missing platform was silently ignored, potentially
loading an empty image:

    $ docker image load -i image.tar --platform=linux/riscv64
    Loaded image: alpine:latest

    $ docker image ls --tree
    IMAGE           ID             DISK USAGE   CONTENT SIZE   USED
    alpine:latest   beefdbd8a1da           0B             0B

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented:
One linting failures;


```
daemon/containerd/image_exporter.go:288: File is not `goimports`-ed (goimports)
			
daemon/containerd/fake_service_test.go:7: File is not `goimports`-ed (goimports)
	"github.com/containerd/containerd/leases"
	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
daemon/containerd/image_exporter_test.go:6: File is not `goimports`-ed (goimports)
	"github.com/containerd/containerd/content"
	"github.com/containerd/containerd/content/local"
```

vvoland commented:
Thanks, fixed! I think I finally got my Goland import settings right, so it should be fine now 😅

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -342,3 +378,62 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+			if pm.Match(imgPlat) {
+				available, err := platformImg.CheckContentAvailable(ctx)
+				if err != nil {
+					return errors.Wrapf(err,
+						"failed to determine image content availability for platform %s",
+						platforms.FormatAll(platform),
+					)
```
And _MAYBE_ if it's not nested, this would fit nicely on a single line.

vvoland replied:
That's still over 120 chars, so I'll keep it as multiline

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -342,3 +378,62 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+
+			if pm.Match(imgPlat) {
+				available, err := platformImg.CheckContentAvailable(ctx)
+				if err != nil {
```
Note to self (and others): I checked  `platformImg.CheckContentAvailable(ctx)`, and it only returns an error on failures, not on "missing" (i.e. `NotFound` errors are not returned)

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -342,3 +378,62 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+// verifyImagesProvidePlatform checks if the requested platform is loaded.
+// If the requested platform is not loaded, it returns an error.
+func (i *ImageService) verifyImagesProvidePlatform(ctx context.Context, imgs []containerdimages.Image, platform ocispec.Platform, pm platforms.Matcher) error {
+	var imageNames []string
```
Perhaps good to name this `missingImageNames` (I had to read the last part of this function twice to get that this slice _NOT_ being empty is the error condition)

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -342,3 +378,62 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+
+	msg := ""
+	switch len(imageNames) {
+	case 0:
+	case 1:
```
It's good to put a comment in empty branches, as they can easily be read as "fall through" (which is the behavior in other languages), e.g.;


```
	case 0:
		// success; none of the images are missing the requested platform.
```

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -342,3 +378,62 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+
+// verifyImagesProvidePlatform checks if the requested platform is loaded.
+// If the requested platform is not loaded, it returns an error.
+func (i *ImageService) verifyImagesProvidePlatform(ctx context.Context, imgs []containerdimages.Image, platform ocispec.Platform, pm platforms.Matcher) error {
```
Wondering; should an empty `imgs` slice also return an error? I _THINK_ currently an empty `imgs` slice also will result in an empty `imageNames` slice ( "no images missing"), and because of that return "no error"

vvoland replied:
Don't think so. Empty `imgs` should probably be handled by the caller of `client.Import`. On the other hand - that would disallow loading "empty" image indexes, which we do for some tests.

vvoland replied:
Ah right, done

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -342,3 +378,62 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+	default:
+		msg = "images [%s] were loaded, but don't provide the requested platform (%s)"
+
+	}
+
+	if msg != "" {
+		return errdefs.NotFound(fmt.Errorf(msg, allImageNames, platformStr))
+	}
```
Perhaps it's more readable to avoid the extra indirect of creating a message only to change it into an error;


```go
	switch len(imageNames) {
	case 0:
		// success; none of the images are missing the requested platform.
	case 1:
		return errdefs.NotFound(fmt.Errorf("image %s was loaded, but doesn't provide the requested platform (%s)", allImageNames, platformStr))
	default:
		return errdefs.NotFound(fmt.Errorf("images [%s] were loaded, but don't provide the requested platform (%s)", allImageNames, platformStr))
	}
```

thaJeztah reviewed (changes requested):
Thanks! Left some questions and some suggestions

vvoland committed:
c8d/test: Add memoryLabelStore

Add an implementation of label store that's needed to create a local
content store for testing purposes.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on daemon/containerd/image_exporter.go:
```diff
@@ -270,12 +275,44 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+		// Verify that the requested platform is available for the loaded images.
+		// While the ideal behavior here would be to verify whether the input
+		// archive actually supplied them, we're not able to determine that
+		// as the imported index is not returned by the import operation.
```
Yeah, definitely. Will create a tracking issue.

vvoland commented on daemon/containerd/image_exporter.go:
```diff
@@ -270,12 +275,44 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
 		}),
 	}
 
+	if platform == nil {
+		// Allow variants to be missing if no specific platform is requested.
+		opts = append(opts, containerd.WithSkipMissing())
+	}
```
The platform matcher will handle that as the non-matching platforms will be filtered out and won't be passed to the handler.
This is only needed for the `platform == nil` case, because the matcher matches **any** platform, but has a `Less` implementation that **prefers** the native one.

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -342,3 +378,61 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+		}
+	}
+
+	allImageNames := strings.Join(missingImages, ", ")
```
Perhaps this var also needs a rename now (`allIMissingImages` or would ... "incomplete images"? be a more appropriate term?)

vvoland replied:
Actually just inlined  that into the fmt.Errorf call (while still keeping it separate from the message itself).
Let me know if that looks fine to you 😅

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -342,3 +378,61 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+
+	msg := ""
+	switch len(missingImages) {
+	case 0:
+		// Success - All images provide the requested platform.
+		return nil
```
As you're switching to an early return, I guess it could be before the `allImageNames` part above 😅

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -342,3 +378,61 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
+		// Success - All images provide the requested platform.
+		return nil
+	case 1:
+		msg = "image %s was loaded, but doesn't provide the requested platform (%s)"
```
I should probably check the actual code, but was the image actually _loaded_ in this case? Will it be garbage-collected immediately after if it doesn't match what we want?

(in that case, considering if `was loaded` would potentially only add confusion, and if a shorter error would work;


```
image %s doesn't provide the requested platform (%s)
```


We could mention "contained in the archive", but not sure if that works;

```
image %s contained in the archive doesn't provide the requested platform (%s)
```

thaJeztah replied:
This sounds like a possible tracking ticket in the making 😂 ❤️

thaJeztah replied:
So I gave this a try with this PR, and while the content is loaded, I _THINK_ it discards the unused ones after? Here's what I did.

I had an alpine image with 2 variants, and exported that one;


```console
$ docker image ls --tree
IMAGE                   ID             DISK USAGE   CONTENT SIZE   USED
alpine:latest           beefdbd8a1da       24.2MB         7.46MB
├─ linux/riscv64        80cde017a105       10.6MB         3.37MB
├─ linux/arm64/v8       9cee2b382fe2       13.6MB         4.09MB
├─ linux/amd64          33735bd63cf8           0B             0B
├─ linux/arm/v6         50f635c8b04d           0B             0B
├─ linux/arm/v7         f2f82d424957           0B             0B
├─ linux/386            b3e87f642f5c           0B             0B
├─ linux/ppc64le        c7a6800e3dc5           0B             0B
└─ linux/s390x          2b5b26e09ca2           0B             0B

$ docker image save -o alpine2.tar alpine:latest
```

I removed the image (daemon now had no images);

```console
$ docker image rm alpine:latest
Untagged: alpine:latest
Deleted: sha256:beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d

$ docker system prune -af
Total reclaimed space: 0B
```

Then loaded the image;

```console
$ docker image load --platform=linux/arm64/v8 --platform=linux/riscv64 -i alpine2.tar
Loaded image: alpine:latest
```

Logs produced as part of this show that content is being removed;

```console
INFO[2024-10-23T07:40:15.747273425Z] API listen on /var/run/docker.sock
DEBU[2024-10-23T07:59:00.084397751Z] handling HEAD request                         method=HEAD module=api request-url=/_ping spanID=7b34e8774792cc0f status=200 traceID=92831a6f26302569d4ea318b0f4c65c4 vars="map[]"
DEBU[2024-10-23T07:59:00.244992084Z] handling POST request                         method=POST module=api request-url="/v1.48/images/load?platform=%7B%22architecture%22%3A%22arm64%22%2C%22os%22%3A%22linux%22%2C%22variant%22%3A%22v8%22%7D&quiet=0" spanID=22be75fe2aca9ff6 status=200 traceID=8e7b378cd832282342c096931cd42c4e vars="map[version:1.48]"
DEBU[2024-10-23T07:59:00.090846709Z] (*service).Write started                      ref=tar-blobs/sha256/2f32f6b11fa159940aadedb7a73a2834f1314ad99e989e38e12dff7ba6575d03 total=1473
DEBU[2024-10-23T07:59:00.093687834Z] (*service).Write started                      ref=tar-blobs/sha256/80cde017a10529a18a7274f70c687bb07c4969980ddfb35a1b921fda3a020e5b total=528
DEBU[2024-10-23T07:59:00.096256418Z] (*service).Write started                      ref=tar-blobs/sha256/8c4a05189a5fd2cf629c25ab8d0831be7156d74b336f129a412933ee78af018c total=3371452
DEBU[2024-10-23T07:59:00.131533543Z] (*service).Write started                      ref=tar-blobs/sha256/9cee2b382fe2412cd77d5d437d15a93da8de373813621f2e4d406e3df0cf0e7c total=528
DEBU[2024-10-23T07:59:00.135708668Z] (*service).Write started                      ref=tar-blobs/sha256/beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d total=1853
DEBU[2024-10-23T07:59:00.139100543Z] (*service).Write started                      ref=tar-blobs/sha256/c157a85ed455142fd79bff5dce951fd5f5b0d0c6e45e6f54cfd0c4e2bdec587b total=1486
DEBU[2024-10-23T07:59:00.141611376Z] (*service).Write started                      ref=tar-blobs/sha256/cf04c63912e16506c4413937c7f4579018e4bb25c272d989789cfba77b12f951 total=4087646
DEBU[2024-10-23T07:59:00.167765126Z] (*service).Write started                      ref=tar-index.json total=441
DEBU[2024-10-23T07:59:00.174570584Z] create image                                  name="docker.io/library/alpine:latest" target="sha256:beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d"
DEBU[2024-10-23T07:59:00.181258376Z] stat snapshot                                 key="sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf"
DEBU[2024-10-23T07:59:00.182770793Z] stat snapshot                                 key="sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf"
DEBU[2024-10-23T07:59:00.183003918Z] prepare snapshot                              key="extract-182858084-qVQF sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf" parent=
DEBU[2024-10-23T07:59:00.185013376Z] using pigz for decompression
DEBU[2024-10-23T07:59:00.234568126Z] diff applied                                  d=49.6165ms digest="sha256:cf04c63912e16506c4413937c7f4579018e4bb25c272d989789cfba77b12f951" media=application/vnd.docker.image.rootfs.diff.tar.gzip size=4087646
DEBU[2024-10-23T07:59:00.235144709Z] commit snapshot                               key="extract-182858084-qVQF sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf" name="sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf"
DEBU[2024-10-23T07:59:00.238797918Z] unpack                                        alreadyUnpacked=false error="<nil>" image="alpine:latest" manifest="sha256:9cee2b382fe2412cd77d5d437d15a93da8de373813621f2e4d406e3df0cf0e7c" spanID=22be75fe2aca9ff6 traceID=8e7b378cd832282342c096931cd42c4e
DEBU[2024-10-23T07:59:00.367122126Z] remove content                                key="sha256:2f32f6b11fa159940aadedb7a73a2834f1314ad99e989e38e12dff7ba6575d03"
DEBU[2024-10-23T07:59:00.367161501Z] remove content                                key="sha256:355eee6af939abf5ba465c9be69c3b725f8d3f19516ca9644cf2a4fb112fd83b"
DEBU[2024-10-23T07:59:00.367171876Z] remove content                                key="sha256:80cde017a10529a18a7274f70c687bb07c4969980ddfb35a1b921fda3a020e5b"
DEBU[2024-10-23T07:59:00.367177876Z] remove content                                key="sha256:8c4a05189a5fd2cf629c25ab8d0831be7156d74b336f129a412933ee78af018c"
DEBU[2024-10-23T07:59:00.367410501Z] schedule content cleanup
DEBU[2024-10-23T07:59:00.367574335Z] removed content                               digest="sha256:2f32f6b11fa159940aadedb7a73a2834f1314ad99e989e38e12dff7ba6575d03"
DEBU[2024-10-23T07:59:00.367631835Z] removed content                               digest="sha256:355eee6af939abf5ba465c9be69c3b725f8d3f19516ca9644cf2a4fb112fd83b"
DEBU[2024-10-23T07:59:00.367656835Z] removed content                               digest="sha256:80cde017a10529a18a7274f70c687bb07c4969980ddfb35a1b921fda3a020e5b"
DEBU[2024-10-23T07:59:00.367862751Z] removed content                               digest="sha256:8c4a05189a5fd2cf629c25ab8d0831be7156d74b336f129a412933ee78af018c"
```

And `docker image ls --tree` only shows one image taking up space:

```console
$ docker image ls --tree
IMAGE                   ID             DISK USAGE   CONTENT SIZE   USED
alpine:latest           beefdbd8a1da       13.6MB         4.09MB
├─ linux/amd64          33735bd63cf8           0B             0B
├─ linux/arm/v6         50f635c8b04d           0B             0B
├─ linux/arm/v7         f2f82d424957           0B             0B
├─ linux/arm64/v8       9cee2b382fe2       13.6MB         4.09MB
├─ linux/386            b3e87f642f5c           0B             0B
├─ linux/ppc64le        c7a6800e3dc5           0B             0B
├─ linux/riscv64        80cde017a105           0B             0B
└─ linux/s390x          2b5b26e09ca2           0B             0B
```

So I _think_ we're good, but yeah, would be nice if we could skip the actual `load` -> `remove` part.

vvoland replied:
In this case it will work correctly - I was talking about a case where the load fails because the archive didn't contain the requested platform. For example in the case above (without this PR):

```bash
$ docker load --platform linux/s390x -i alpine2.tar 
Loaded image: alpine:latest

$ docker images --tree
WARNING: This is an experimental feature. The output may change and shouldn't be depended on.

IMAGE           ID             DISK USAGE   CONTENT SIZE   IN USE
alpine:latest   beefdbd8a1da           0B             0B        
```

The image is empty, because only the index got loaded and the image was already created by `client.Import`.

thaJeztah replied:
Ah! Alright, so yes, with this PR that works correctly;

```bash
docker load --platform linux/s390x -i alpine2.tar
requested platform (linux/s390x) found, but some content is missing: NotFound: content digest sha256:2b5b26e09ca2856f50ac88312348d26c1ac4b8af1df9f580e5cf465fd76e3d4d: not found

docker image ls --tree
IMAGE   ID             DISK USAGE   CONTENT SIZE   IN USE

docker load -i alpine2.tar
Loaded image: alpine:latest

docker image ls --tree
IMAGE                   ID             DISK USAGE   CONTENT SIZE   IN USE
alpine:latest           beefdbd8a1da       16.9MB         7.46MB
├─ linux/amd64          33735bd63cf8           0B             0B
├─ linux/arm/v6         50f635c8b04d           0B             0B
├─ linux/arm/v7         f2f82d424957           0B             0B
├─ linux/arm64/v8       9cee2b382fe2       13.6MB         4.09MB
├─ linux/386            b3e87f642f5c           0B             0B
├─ linux/ppc64le        c7a6800e3dc5           0B             0B
├─ linux/riscv64        80cde017a105       3.37MB         3.37MB
└─ linux/s390x          2b5b26e09ca2           0B             0B
```

thaJeztah replied:
Tracking ticket;

- https://github.com/moby/moby/issues/48762

thaJeztah commented on daemon/containerd/image_load_test.go:
```diff
@@ -0,0 +1,111 @@
+
+		err = tryLoad(ctx, t, imgDataDir, linuxArm64)
+		assert.Check(t, is.ErrorContains(err, "doesn't provide the requested platform (linux/arm64)"))
+		assert.Check(t, errdefs.IsNotFound(err))
```
Oh! Same comment here as on the other PR; https://github.com/moby/moby/pull/48722#discussion_r1810554194

> Perhaps use `is.ErrorType` - It's currently marked "deprecated", but will be un-deprecated in the next release; [gotestyourself/gotest.tools#286](https://github.com/gotestyourself/gotest.tools/pull/286)
> 
> ```go
> assert.Check(t, is.ErrorType(err, errdefs.IsNotFound))
> ```
> 
> Alternattively, add a custom comment to show the actual error-type and message;
> 
> ```go
> assert.Check(t, errdefs.IsNotFound(err), "expected  a ErrNotFound, got %[1]v (%[1]T)", err)
> ```

vvoland replied:
Ah, good to heart that it will be undeprecated! ❤️

thaJeztah replied:
Yeah, I understand the motivation (i.e., trying to keep `gotest.tools` as minimal as possible; don't duplicate assertions where there's no need), but the `is.ErrorType` was really much more convenient for our cases 😂

vvoland commented on daemon/containerd/fake_service_test.go:
```diff
@@ -4,21 +4,22 @@ import (
 	"fmt"
-	"github.com/containerd/containerd/leases"
-	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"os"
```
OOF, thanks for noticing!

thaJeztah commented on daemon/containerd/image_exporter.go:
```diff
@@ -253,11 +253,16 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
 
 	pm := i.matchRequestedOrDefault(platforms.OnlyStrict, platform)
 
+	// Unpack only an image of the host platform
+	unpackPm := i.hostPlatformMatcher()
+	// If a load of specific platform is requested, unpack it
+	if platform != nil {
+		unpackPm = pm
+	}
+
```
Last nit; this one is not used until much further below; perhaps move it until after we did the  ` i.verifyImagesProvidePlatform` call

thaJeztah reviewed (approved):
LGTM


one thing I did _not_ verify is what we do with attestations; do we preserve those when saving (and loading)?

thaJeztah commented:
And flakiness on Windows; I'm starting to wonder if we have some regression or if something changed on the GitHub runners; seeing `has active endpoints` errors come up various times;


```
=== FAIL: github.com/docker/docker/integration/networking TestNatNetworkICC/User_defined_nat_network (8.50s)
    nat_windows_test.go:62: assertion failed: error is not nil: Post "http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.48/containers/388543970cc75202326c968cdb2544f94cf4a096072a90e20729e7182f4ab608/start": context deadline exceeded
    panic.go:626: assertion failed: error is not nil: Error response from daemon: error while removing network: network mynat id 9f8696976cb154606f63265e3f6e812855a1b9cb87f9fe268f64343240cc0485 has active endpoints
```

vvoland commented:
> one thing I did not verify is what we do with attestations; do we preserve those when saving (and loading)?

Currently we only preserve them if no `platform` was selected. Technically we _could_ also load a corresponding attestation, but I don't think we can do that with `client.Import` - we'd need a lower level code for this.

 
> Do we need API changes to make that possible, or would the API already be ablee to handle multiple? 

We can support multiple platforms with the current API via multiple `platform` query fields, like:
```
localhost/images/load?platform={"architecture":"arm64","os":"linux","variant":"v8"}&platform={"architecture":"arm","os":"linux","variant":"v5"}`
```

thaJeztah commented:
> We can support multiple platforms with the current API via multiple platform query fields, like:

Great! I should give that a try. We must change the Client and types to prepare for that though (we didn't ship this change yet, so we can still do before v28.0); I created a tracking ticket;


- https://github.com/moby/moby/issues/48759

thaJeztah commented:
> Currently we only preserve them if no platform was selected. Technically we could also load a corresponding attestation, but I don't think we can do that with `client.Import` - we'd need a lower level code for this.

created a tracking ticket for that as well;

- https://github.com/moby/moby/issues/48761

laurazard reviewed (approved):
LGTM

thaJeztah commented:
Tracking tickets created for various things; let's bring this one in! Thanks @vvoland !!

thaJeztah commented:
Removed the cherry-pick label, because this feature / flag is not yet in v27.x 🙈
